File: 683/683_version2.py
class Solution:
    def kEmptySlots(self, flowers: List[int], k: int) -> int:
        # version 2: Bucket
        size = len(flowers)
        if size==0 | k>=size: # corner case test
            return -1
        k+=1
        bucket_size = (size+k-1)//k # calculate the number of bucket, this is to compress the status
        low_bucket = [float("inf")]*bucket_size
        high_bucket = [-float("inf")]*bucket_size
        for i in range(size):
            flower_pos = flowers[i]
            bucket_pos = flowers[i]//k
            day = i+1
            if flower_pos < low_bucket[bucket_pos]:
                low_bucket[bucket_pos] = flower_pos
                if bucket_pos > 0 and high_bucket[bucket_pos-1]==flower_pos-k:
                    return day
            if flower_pos > high_bucket[bucket_pos]:
                high_bucket[bucket_pos] = flower_pos
                if bucket_pos < bucket_size-1 and low_bucket[bucket_pos+1]==flower_pos+k:
                    return day
        return -1

File: 683/683_version2.java
class Solution {
    public int kEmptySlots(int[] flowers,int k){
        /* version 2: bucket */
        int n = flowers.length, p=k+1,bs=n/p + 1;
        int[] low_b = new int[bs];
        int[] high_b = new int[bs];
        Arrays.fill(low_b,Integer.MAX_VALUE);
        Arrays.fill(high_b,Integer.MIN_VALUE);
        for(int i=0;i<n;i++){
            int cp = flowers[i];
            int bp = cp/p;
            int day = i+1;
            if(cp < low_b[bp]){
                low_b[bp] = cp;
                if(bp-1>=0 && high_b[bp-1]==cp-p) return day;
            }
            if(cp > high_b[bp]){
                high_b[bp] = cp;
                if(bp+1 < bs && low_b[bp+1]==cp+p) return day;
            }
        }
        return -1;
    }
}